staging/src/k8s.io/apimachinery/pkg/util/waitgroup/optionalwaitgroup.go

```diff
@@ -0,0 +1,55 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package waitgroup
+
+import (
+	"sync"
+)
+
+// OptionalWaitGroup is a WaitGroup or nothing.  Use it to track when
+// goroutines start or stop or begin or end waiting if the pointer is
+// not nil.  This abstraction is intended to make it easier to write a
+// package that can be unit tested by a function that advances a fake
+// clock once all the available work is done.  Real clients set the
+// pointer to nil because they do not need this tracking.
```

Comment:
placing test code into mainline paths worries me.  I'll see the actual usages further on.

Reply:
OptionalWaitGroup is not just for test. It is used in the mainline code.  It's whole purpose is to streamline handling of a maybe-nil pointer to a WaitGroup.

Reply:
> OptionalWaitGroup is not just for test. It is used in the mainline code. It's whole purpose is to streamline handling of a maybe-nil pointer to a WaitGroup.

I'm still working my way down, but it's pretty unusual to only sometimes need a waitgroup.  @lavalamp calling this out to see if you're comfortable with usage.

Reply:
This is perhaps an unusual use of WaitGroup but I think it is valid.  This code (the mainline code, not the test code) needs the ability to keep track of whether there are goroutines active IF its client wants that.  Keeping track of active goroutines is exactly what WaitGroup is about.  Using a maybe-nil pointer to one gives the optionality.